File: app/src/main/java/me/echeung/moemoekyun/ui/base/BaseActivity.java
package me.echeung.moemoekyun.ui.base;

import android.content.Context;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.view.MenuItem;

import androidx.appcompat.app.AppCompatActivity;
import me.echeung.moemoekyun.App;
import me.echeung.moemoekyun.util.PreferenceUtil;
import me.echeung.moemoekyun.util.system.LocaleUtil;
import me.echeung.moemoekyun.util.system.ThemeUtil;

public abstract class BaseActivity extends AppCompatActivity implements SharedPreferences.OnSharedPreferenceChangeListener {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        ThemeUtil.INSTANCE.setTheme(this);
        LocaleUtil.INSTANCE.setTitle(this);

        ThemeUtil.INSTANCE.colorNavigationBar(this);

        App.Companion.getPreferenceUtil().registerListener(this);

        super.onCreate(savedInstanceState);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (item.getItemId() == android.R.id.home) {
            finishAfterTransition();
            return true;
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    protected void onDestroy() {
        App.Companion.getPreferenceUtil().unregisterListener(this);

        super.onDestroy();
    }

    @Override
    protected void attachBaseContext(Context base) {
        super.attachBaseContext(LocaleUtil.INSTANCE.setLocale(base));
    }

    @Override
    public void onSharedPreferenceChanged(SharedPreferences sharedPreferences, String key) {
        switch (key) {
            case PreferenceUtil.PREF_COLOR_NAVBAR:
                ThemeUtil.INSTANCE.colorNavigationBar(this);
                break;
        }
    }

}

File: app/src/main/kotlin/me/echeung/moemoekyun/ui/base/BaseActivity.kt
package me.echeung.moemoekyun.ui.base

import android.content.Context
import android.os.Bundle
import android.view.MenuItem
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import me.echeung.moemoekyun.R
import me.echeung.moemoekyun.util.system.LocaleUtil
import org.koin.android.ext.android.inject

abstract class BaseActivity : AppCompatActivity() {

    private val localeUtil: LocaleUtil by inject()

    protected var appbar: Toolbar? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        localeUtil.setTitle(this)

        super.onCreate(savedInstanceState)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == android.R.id.home) {
            finishAfterTransition()
            return true
        }
        return super.onOptionsItemSelected(item)
    }

    override fun attachBaseContext(base: Context) {
        super.attachBaseContext(localeUtil.setLocale(base))
    }

    open fun initAppbar() {
        appbar = findViewById(R.id.appbar)
        appbar?.let {
            setSupportActionBar(it)
            supportActionBar?.setDisplayHomeAsUpEnabled(true)
            supportActionBar?.setDisplayShowTitleEnabled(true)
        }
    }
}
